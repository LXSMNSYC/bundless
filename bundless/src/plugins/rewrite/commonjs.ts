import { isPlainObject } from 'lodash'
import path from 'path'
import fs from 'fs-extra'
import { COMMONJS_ANALYSIS_PATH, WEB_MODULES_PATH } from '../../constants'
import slash from 'slash'
import { ImportDeclaration } from '@babel/types'
import { parse } from '../../utils'
import { makeLegalIdentifier } from '@rollup/pluginutils'
import { osAgnosticPath } from '../../prebundle/support'
import memoize from 'micro-memoize'
import { onResolveLock } from '../../serve'
import { logger } from '../../logger'

export interface OptimizeAnalysisResult {
    isCommonjs: { [name: string]: true }
}

/**
 * read analysis result from optimize step
 * If we can't find analysis result, return null
 * (maybe because user set optimizeDeps.auto to false)
 */
export const getAnalysis = memoize(function getAnalysis(
    root: string,
): OptimizeAnalysisResult | null {
    let analysis: OptimizeAnalysisResult | null
    try {
        analysis = fs.readJsonSync(
            path.resolve(root, WEB_MODULES_PATH, COMMONJS_ANALYSIS_PATH),
        )
    } catch (error) {
        logger.debug(
            `Cannot find commonjs analysis at ${path.resolve(
                root,
                WEB_MODULES_PATH,
                COMMONJS_ANALYSIS_PATH,
            )}`,
        )
        analysis = null
    }
    if (analysis && !isPlainObject(analysis.isCommonjs)) {
        throw new Error(`invalid ${COMMONJS_ANALYSIS_PATH}`)
    }
    logger.debug(
        `Got new commonjs analysis: ${JSON.stringify(
            analysis?.isCommonjs,
            null,
            4,
        )}`,
    )
    return analysis
})

export function clearCommonjsAnalysisCache() {
    logger.debug(`Invalidating commonjs cache`)
    getAnalysis.cache.keys.length = 0
    getAnalysis.cache.values.length = 0
}

export function isOptimizedCjs(root: string, filename: string) {
    if (!onResolveLock.isReady) {
        throw new Error(
            `Cannot call isOptimizedCjs when onResolveLock is locked!`,
        )
    }
    const analysis = getAnalysis(root)
    if (!analysis) {
        return false
    }
    const isCommonjs = !!analysis.isCommonjs[osAgnosticPath(filename, root)]
    return isCommonjs
}

type ImportNameSpecifier = { importedName: string; localName: string }

// todo if module has __esModule and there is only a default import, transform to .default, -> const imported = realImport.__esModule ? realImport.default : realImport
export function transformCjsImport(
    exp: string,
    id: string,
    resolvedPath: string,
    importIndex: number,
): string {
    const ast = parse(exp)[0] as ImportDeclaration
    const importNames: ImportNameSpecifier[] = []

    ast.specifiers.forEach((obj) => {
        if (
            obj.type === 'ImportSpecifier' &&
            obj.imported.type === 'Identifier'
        ) {
            const importedName = obj.imported.name
            const localName = obj.local.name
            importNames.push({ importedName, localName })
        } else if (obj.type === 'ImportDefaultSpecifier') {
            importNames.push({
                importedName: 'default',
                localName: obj.local.name,
            })
        } else if (obj.type === 'ImportNamespaceSpecifier') {
            importNames.push({ importedName: '*', localName: obj.local.name })
        }
    })

    return generateCjsImport(importNames, id, resolvedPath, importIndex)
}

function generateCjsImport(
    importNames: ImportNameSpecifier[],
    id: string,
    resolvedPath: string,
    importIndex: number,
): string {
    // If there is multiple import for same id in one file,
    // importIndex will prevent the cjsModuleName to be duplicate
    const cjsModuleName = makeLegalIdentifier(`${id}_cjsImport${importIndex}`)
    const lines: string[] = [`import ${cjsModuleName} from "${resolvedPath}";`]
    importNames.forEach(({ importedName, localName }) => {
        if (importedName === '*' || importedName === 'default') {
            lines.push(
                `const ${localName} = ${cjsModuleName} && ${cjsModuleName}.__esModule ? ${cjsModuleName}.default : ${cjsModuleName};`,
            )
        } else {
            lines.push(
                `const ${localName} = ${cjsModuleName}["${importedName}"];`,
            )
        }
    })
    return lines.join('\n')
}
