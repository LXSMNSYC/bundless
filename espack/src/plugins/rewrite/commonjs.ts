import { isPlainObject } from 'lodash'
import path from 'path'
import fs from 'fs-extra'
import { COMMONJS_ANALYSIS_PATH, WEB_MODULES_PATH } from '../../constants'
import slash from 'slash'
import { ImportDeclaration } from '@babel/types'
import { parse } from '../../utils'
import { makeLegalIdentifier } from '@rollup/pluginutils'

const analysisCache = new Map<string, OptimizeAnalysisResult | null>()

export interface OptimizeAnalysisResult {
    isCommonjs: { [name: string]: true }
}

/**
 * read analysis result from optimize step
 * If we can't find analysis result, return null
 * (maybe because user set optimizeDeps.auto to false)
 */
function getAnalysis(root: string): OptimizeAnalysisResult | null {
    if (analysisCache.has(root)) return analysisCache.get(root)!
    let analysis: OptimizeAnalysisResult | null
    try {
        const cacheDir = path.resolve(root, WEB_MODULES_PATH)
        analysis = fs.readJsonSync(path.join(cacheDir!, COMMONJS_ANALYSIS_PATH))
    } catch (error) {
        analysis = null
    }
    if (analysis && !isPlainObject(analysis.isCommonjs)) {
        throw new Error(`[vite] invalid ${COMMONJS_ANALYSIS_PATH}`)
    }
    analysisCache.set(root, analysis)
    return analysis
}

export function isOptimizedCjs(root: string, filename: string) {
    console.log(`isOptimizedCjs ${filename}`)
    const analysis = getAnalysis(root)
    if (!analysis) return false
    return !!analysis.isCommonjs[slash(path.relative(root, filename))]
}

type ImportNameSpecifier = { importedName: string; localName: string }

export function transformCjsImport(
    exp: string,
    id: string,
    resolvedPath: string,
    importIndex: number,
): string {
    const ast = parse(exp)[0] as ImportDeclaration
    const importNames: ImportNameSpecifier[] = []

    ast.specifiers.forEach((obj) => {
        if (
            obj.type === 'ImportSpecifier' &&
            obj.imported.type === 'Identifier'
        ) {
            const importedName = obj.imported.name
            const localName = obj.local.name
            importNames.push({ importedName, localName })
        } else if (obj.type === 'ImportDefaultSpecifier') {
            importNames.push({
                importedName: 'default',
                localName: obj.local.name,
            })
        } else if (obj.type === 'ImportNamespaceSpecifier') {
            importNames.push({ importedName: '*', localName: obj.local.name })
        }
    })

    return generateCjsImport(importNames, id, resolvedPath, importIndex)
}

function generateCjsImport(
    importNames: ImportNameSpecifier[],
    id: string,
    resolvedPath: string,
    importIndex: number,
): string {
    // If there is multiple import for same id in one file,
    // importIndex will prevent the cjsModuleName to be duplicate
    const cjsModuleName = makeLegalIdentifier(
        `$viteCjsImport${importIndex}_${id}`,
    )
    const lines: string[] = [`import ${cjsModuleName} from "${resolvedPath}";`]
    importNames.forEach(({ importedName, localName }) => {
        if (importedName === '*' || importedName === 'default') {
            lines.push(`const ${localName} = ${cjsModuleName};`)
        } else {
            lines.push(
                `const ${localName} = ${cjsModuleName}["${importedName}"];`,
            )
        }
    })
    return lines.join('\n')
}
