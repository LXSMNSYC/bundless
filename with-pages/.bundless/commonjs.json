{
    "isCommonjs": {
        ".bundless/web_modules/index-V3344356.js": true,
        ".bundless/web_modules/index-3RER33FG.js": true
    }
}