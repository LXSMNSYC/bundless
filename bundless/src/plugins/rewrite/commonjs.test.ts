import { transformCjsImport } from './commonjs'

describe('rewrite commonjs imports', () => {
    const cases = [
        `import React from 'react'`,
        `import * as React from 'react'`,
        `import React, { useState } from 'react'`,
        `import { useState } from 'react'`,
        `import { useState, useEffect } from 'react'`,
        `import { useState as something } from 'react'`,
        `import { useState as something, useEffect } from 'react'`,
        `import { useState as something, useEffect as alias } from 'react'`,
        `import { default as Default } from 'react'`,
        `import { default as Default, useEffect } from 'react'`,
    ]
    for (let [i, testCase] of cases.entries()) {
        test(`${i} "${testCase}"`, () => {
            const res = transformCjsImport(testCase, 'react', 'react', 0)
            expect(res).toMatchSnapshot()
        })
    }
})
