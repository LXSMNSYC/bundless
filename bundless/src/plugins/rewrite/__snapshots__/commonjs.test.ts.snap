// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rewrite commonjs imports 0 "import React from 'react'" 1`] = `"import react_cjsImport0 from \\"react\\"; const React = react_cjsImport0 && react_cjsImport0.__esModule ? react_cjsImport0.default : react_cjsImport0;"`;

exports[`rewrite commonjs imports 1 "import * as React from 'react'" 1`] = `"import react_cjsImport0 from \\"react\\"; const React = {default: react_cjsImport0, ...(typeof react_cjsImport0 === 'object' && react_cjsImport0)};"`;

exports[`rewrite commonjs imports 2 "import React, { useState } from 'react'" 1`] = `"import react_cjsImport0 from \\"react\\"; const React = react_cjsImport0 && react_cjsImport0.__esModule ? react_cjsImport0.default : react_cjsImport0; const useState = react_cjsImport0[\\"useState\\"];"`;

exports[`rewrite commonjs imports 3 "import { useState } from 'react'" 1`] = `"import react_cjsImport0 from \\"react\\"; const useState = react_cjsImport0[\\"useState\\"];"`;

exports[`rewrite commonjs imports 4 "import { useState, useEffect } from 'react'" 1`] = `"import react_cjsImport0 from \\"react\\"; const useState = react_cjsImport0[\\"useState\\"]; const useEffect = react_cjsImport0[\\"useEffect\\"];"`;

exports[`rewrite commonjs imports 5 "import { useState as something } from 'react'" 1`] = `"import react_cjsImport0 from \\"react\\"; const something = react_cjsImport0[\\"useState\\"];"`;

exports[`rewrite commonjs imports 6 "import { useState as something, useEffect } from 'react'" 1`] = `"import react_cjsImport0 from \\"react\\"; const something = react_cjsImport0[\\"useState\\"]; const useEffect = react_cjsImport0[\\"useEffect\\"];"`;

exports[`rewrite commonjs imports 7 "import { useState as something, useEffect as alias } from 'react'" 1`] = `"import react_cjsImport0 from \\"react\\"; const something = react_cjsImport0[\\"useState\\"]; const alias = react_cjsImport0[\\"useEffect\\"];"`;

exports[`rewrite commonjs imports 8 "import { default as Default } from 'react'" 1`] = `"import react_cjsImport0 from \\"react\\"; const Default = react_cjsImport0 && react_cjsImport0.__esModule ? react_cjsImport0.default : react_cjsImport0;"`;

exports[`rewrite commonjs imports 9 "import { default as Default, useEffect } from 'react'" 1`] = `"import react_cjsImport0 from \\"react\\"; const Default = react_cjsImport0 && react_cjsImport0.__esModule ? react_cjsImport0.default : react_cjsImport0; const useEffect = react_cjsImport0[\\"useEffect\\"];"`;
